text/0000-atcompoundkey-error-path-optimization.md

```diff
@@ -0,0 +1,197 @@
+* If the BundleManifest class is modified to handle "null" values in the manifest, this API has to change.  
+
+## Option 2 - Modify the current API to return an Any object by value
+Return an empty `Any` object if the key is not found
+
+```
+Any AtCompoundKey(const std::string& key);
+```
+
+### Pros
+* Addresses the performance problem. The happy path and the error path have similar performance. 
+
+###Cons
+* Returning by value results in copies of the returned object which could be costly, specially if this API is used to retrieve intermediate maps of larger size.
+* It is not clear how a null value from a manifest file is treated. specifying "null" is legal in JSON. Current implementation of bundle metadata parser ignores the entries with null value. However, we cannot assume this will not change in the future.
+* The information regarding he error conditions is lost. However, do the users really care about the information is a puzzle.
+
+## Option 2.1 (Preferred) - Add an overload with a default value parameter 
+Add an overload with the following signature:
+
+```
+Any AtCompoundKey(const std::string& key, Any defaultvalue) const;
+```
+
+Return the user provided default value if the key is not found in the map.
+
+### Pros
+* Addresses the performance problem. The happy path and the error path have similar performance. 
+* Backwards compatible. Since this API has different number of arguments, this API can be added as an overload set. This allows us to let users choose the right API based on their performance and error handling requirements. Also, lets us deprecate the old one if no use-cases exist.
+
+### Cons
+* User has to provide an extra argument for the default value.
+* The information regarding the error conditions is lost. However, if the users care about the error conditions, they can use the overload version that throws.
+
+## Option 3 - Expose the bundle's metadata as a third party JSON object.
+Remove the `AtCompoundKey` API and expose the bundle's metadata in the form of an object of a JSON parsing library. E.g, rapidjson::Document from the RapidJson library
+
+```
+rapidjson::Document Bundle::GetHeaders();
+```
+
+### Pros
+* Users can use the JSON object to write their custom parsing functions or pass around JSON objects in their code 
+
+### Cons
+* Users have to parse through the JSON tree to retrieve the values. 
+* The team currently using AtCompoundKey does not like this option. They do not want to deal with the any_cast at each level of the map
+
+## Option 4
+
+Add a "KeyExists" API which returns a bool to signify the existence of the entry.
+
+```
+if(map.KeyExists("com.foo.bar"))
+{
+  auto val = map.AtCompoundKey("com.foo.bar");
+}
+```
+
+### Pros
+* Similar to the STL map's "count" API
+* Backwards compatible
+
+### Cons
+* Users have to call this API before calling the existing AtCompoundKey API. The resulting code would require parsing the dot notation key twice which may not be desirable.
```

Comment:
There seems to be two potential cons here: 
1. the extra API call (to ```KeyExists```) clients need to make, which is not as simple as having to make one API call.
2. a possible performance hit from parsing a string and traversing the map twice.